network/network_interface.ml

```diff
@@ -25,299 +27,547 @@ let uri () = "file:" ^ !default_path
-	mtu = 1500;
-	ethtool_settings = [];
-	ethtool_offload = ["lro", "off"];
-	persistent_i = false;
-}
-let default_bridge = {
-	ports = [];
-	vlan = None;
-	bridge_mac = None;
-	igmp_snooping = None;
-	other_config = [];
-	persistent_b = false;
-}
-let default_port = {
-	interfaces = [];
-	bond_properties = [];
-	bond_mac = None;
-	kind = Basic;
-}
-let default_config = {
-	interface_config = [];
-	bridge_config = [];
-	gateway_interface = None;
-	dns_interface = None
-}
-
-(** {2 RPC functions} *)
-
-let interface_config_t_add_defaults rpc =
-	Rpc.struct_extend rpc (rpc_of_interface_config_t default_interface)
-
-let port_config_t_add_defaults rpc =
-	Rpc.struct_extend rpc (rpc_of_port_config_t default_port)
-
-let bridge_config_t_add_defaults rpc =
-	(* This needs some special treatment, because bridge_config_t contains a list of port_config_t records
-	 * that may need to have defaults inserted. Rpc.struct_extend does not currently support this. *)
-	let open Rpc in
-	let rpc' = Rpc.struct_extend rpc (rpc_of_bridge_config_t default_bridge) in
-	match rpc' with
-	| Dict r ->
-		Dict (List.map (fun (k, v) ->
-			match k, v with
-			| "ports", Dict v' ->
-				k, Dict (List.map (fun (name, config) -> name, port_config_t_add_defaults config) v')
-			| x -> x
-		) r)
-	| x -> x
-
-let config_t_add_defaults rpc =
-	(* This needs some special treatment, because config_t contains lists of bridge_config_t and
-	 * interface_config_t records that may need to have defaults inserted. Rpc.struct_extend does
-	 * not currently support this. *)
-	let open Rpc in
-	let rpc' = Rpc.struct_extend rpc (rpc_of_config_t default_config) in
-	match rpc' with
-	| Dict r ->
-		Dict (List.map (fun (k, v) ->
-			match k, v with
-			| "bridge_config", Dict v' ->
-				k, Dict (List.map (fun (name, config) -> name, bridge_config_t_add_defaults config) v')
-			| "interface_config", Dict v' ->
-				k, Dict (List.map (fun (name, config) -> name, interface_config_t_add_defaults config) v')
-			| x -> x
-		) r)
-	| x -> x
-
-let interface_config_t_of_rpc rpc = rpc |> interface_config_t_add_defaults |> interface_config_t_of_rpc
-let port_config_t_of_rpc rpc =      rpc |> port_config_t_add_defaults      |> port_config_t_of_rpc
-let bridge_config_t_of_rpc rpc =    rpc |> bridge_config_t_add_defaults    |> bridge_config_t_of_rpc
-let config_t_of_rpc rpc =           rpc |> config_t_add_defaults           |> config_t_of_rpc
+  interface_config: (iface * interface_config_t) list [@default []];
+  bridge_config: (bridge * bridge_config_t) list [@default []];
+  gateway_interface: iface option [@default None];
+  dns_interface: iface option [@default None];
+} [@@deriving rpcty]
 
 (** {2 Configuration manipulation} *)
 
-let empty_config = default_config
-
 let get_config config default name =
-	if List.mem_assoc name config = false then
-		default
-	else
-		List.assoc name config
+  if List.mem_assoc name config = false then
+    default
+  else
+    List.assoc name config
 
```

Comment:
This could be simplified like so, though I'm not sure which style is preferred in this repo:
```
try
  List.assoc name config
with _ -> default
```

Reply:
I prefer keep it as is.

Reply:
If you want to keep it as it is, please do

```ocaml
if not List.mem_assoc name config then default
else List.assoc name config
```

I personally also prefer the @minishrink one that avoids a lookup.